proposals/NNNN-mixins.md

```diff
@@ -0,0 +1,501 @@
+
+struct Enemy: MovingObject, AnimatedObject { /*...*/ }
+struct Arrow: MovingObject { /*...*/ }
+```
+
+### Example with logging
+
+`LoggingObject` is not a class of objects in the original meaning. It is more like a feature, or behavior of a data type. Using mixins, we can free the road for other parent classes (or other mixins) in the hierarchy:
+
+```swift
+class Logger { /*...*/ }
+
+mixin LoggingObject {
+  var logger: Logger
+  init(logger: Logger) { self.logger = logger }
+}
+
+class GenericComputer {
+  var result: String? = nil
+  init(options: Options) { /*...*/ }
+  // ...
+}
+
+class SomeComputer: GenericComputer, LoggingObject {
+  init(logger: Logger, options: Options) {
+    LoggingObject.super.init(logger: logger)
+    super.init(options: options)
+  }
+  
+  override func compute() {
+    logger.log("Starting", 1)
+    if someCondition {
+      logger.log("Error", 3)
+    }
+    logger.log("Ending", 1)
+  }
+}
+```
+
+## Detailed design
+
+### Mixin members 
+
+Property requirements:
+
+```swift
+var name: Type { get set }
+```
+
+Property definitions:
+
+```swift
+var name: Type
+```
+
+Method requirements:
+
+```swift
+func name(arg: Type) -> Result
+```
+
+Method definitions:
+
+```swift
+func name(arg: Type) -> Result { }
+```
+
+There is no notion of default methods in mixins. Methods of mixins require `override` to redefine them in types and can be `final`. 
+
+Mixins consider themselves as value types, as protocols and structs. Every member and requirement, which needs to mutate `self`, must be declared as `mutating`.
+
+Subscript requirements and definitions.
+
+Initializer requirements:
+
+```swift
+init(name: Type)
+```
+
+Initializer definitions:
+
+```swift
+init(name: Type) { }
+```
+
+Properties can be immediately initialized: `var name: Type = initializer`
+
+These initialized properties are prepended to all initializers. If no initializers are defined, an implicit default initializer is created.
+
+### Initializers
+
+Mixins can have implemented initializers. They act like initializers of classes. `super` initializer calls are prefixed with parent protocol name for disambiguation purposes. `super` initializer must be explicitly called in initializer of subtype. The syntax for that is `SuperMixinName.super.init()`:
+
+```swift
+mixin A {
+  init(newPath: String) { }
+}
+
+mixin B: A {
+  override init(newPath: String) { }
+}
+```
+
+Default `super` initializers also need to be called explicitly. In multiple inheritance, it is a responsibility of subtype to ensure that initialization order matches semantics of mixins it mixes in:
+
+```swift
+mixin A { }
+mixin B { }
+
+mixin C: A, B {
+  init() {
+    A.super.init()
+    B.super.init()
+  }
+}
+```
+
+Note that `super` syntax can also be used for calling supertype members, as in classes.
+
+### Mixin inheritance
+
+If a single mixin is mixed-in using different paths, then only one copy of it will be kept in the final struct/class. Example:
```

Comment:
I believe this is a bad idea. I believe the developer should **always** have to explicitly define whether or not the base mixin should be copied twice or if only one copy should be used.

Here's an example where you might only want 1 copy of the base mixin:

```swift
mixin Mover {
    var xPosition: Int
}

mixin Jumper: Mover { ... }
mixin Walker: Mover { ... }

class Human: Walker, Jumper {
    ...
}
```

It is clear that even though a human can be both a walker and a jumper, a human can only be in one place at a time, so it should only inherit one copy of the `xPosition: Int` of Mover. This means that both `Walker` and `Jumper` will work with the same variable `xPosition`.

Here's a case where you might want 2 copies of the base mixin (shamelessly stolen from the mailing list):

```swift
mixin UniversityPerson {
    var computerAccount: Account
}
mixin Teacher: UniversityPerson { ... }
mixin Student: UniversityPerson { ... }

class TeachingAssistant: Teacher, Student { 
    ... 
}
```

Here a TeachingAssistant might actually have two different computerAccounts. One for his studying activities and one for his teaching activities. So in this case the developer would need to rename one or both of the versions of the inherited variable.

There's no reliable way to make assumptions about when things should be copied once or multiple times, which is why I propose to force the developer to explicitly define it. Here's a possible syntax we could use for that:

```swift
class TeachingAssistant: Teacher[computerAccount as teacherAccount],
                         Student[computerAccount as studentAccount] { 
    // We explicitly conveyed that we need
    // two different copies of computerAccount
}

class Human: Walker[xPosition],
             Jumper[xPosition] {
    // We explicitly conveyed that we want walker and jumper
    // to work with the same copy of the variable xPosition
    // [xPosition] is just syntactic sugar for [xPosition as xPosition]
}
```

And we could easily mix and match that like so:

```swift
class C: M1[someVar],
         M2[someVar],
         M3[someVar as someVar2],
         M4[someVar as someVar2],
         M5[someVar as someVar3] {
    // M1 and M2 use the same copy of someVar
    // M3 and M4 also use the same copy,
    // but different from M1 and M2
    // and M5 uses an entirely different copy again
}
```